### DISCUSSION TITLE
How to get edges between clusters using "ClusterData"?

### DISCUSSION THREAD

**Discussion by ObsisMc** [pyg-team/pytorch_geometric#6885] [Q&A] [answered]
Are there any methods can I use to get edges between different clusters after using `ClusterData`

**Comment by wsad1** [comment, marked as answer]
That class needs better documentation. 
`ClusterData` contains `perm` and `partptr`. `perm` is the new permutation of nodes and `partptr` is of shape `num_parts+1` and assigns node slices to a partition.  You should be able to use these two to find edges that go across clusters.

**Reply by ObsisMc** [reply]
OK, I understand meanings of these attributes. In fact, I hope there will be a function to get those edges (maybe can be gotten during clustering? I am not familiar with the algorithm) since I think my implementation will be slow 😂.

By the way, `adj` of `ClusterData` isn't the adjacency matrix of clustered data and I cannot use it to find edges between clusters, right?

**Reply by rusty1s** [reply]
`adj` is the permuted adjacency matrix, grouped according to the partitions. What do you mean with finding edges between clusters?

**Reply by ObsisMc** [reply]
For example
![Screenshot_20230310_160005](https://user-images.githubusercontent.com/71996633/224257762-152e5f8a-30c6-49f0-bca7-58505bda5b9d.jpg)

Triangles and circles are nodes, and they are in different clusters. Green lines are edges in different clusters and red lines are edges connecting the two clusters. 

Then, I want to get red lines.

**Reply by rusty1s** [reply]
If you know the cluster, you can do something like `adj[partptr[i]:partptr[i+1], partptr[j]:partptr[j + 1]]`.

**Reply by ObsisMc** [reply]
Sorry but I cannot understand the `adj` and don't know how to use it.
In class `ClusterData`, `adj` is defined and operated here:
```python
            adj = SparseTensor(
                row=data.edge_index[0], col=data.edge_index[1],
                value=torch.arange(E, device=data.edge_index.device),
                sparse_sizes=(N, N))
            adj, partptr, perm = adj.partition(num_parts, recursive)
```
Do meanings of `row` and `col` attributes in `adj` change after partition? It seems that the `adj` isn't in COO format since I find it contains some edges which don't exist in original graph.

**Reply by rusty1s** [reply]
Yes, `row` and `col` are permuted after the `adj.partition` call, so the indexing change such that the first n rows and n cols refer to the indices of the first partition. You can use the indexing logic as described above to get inter- and intra cluster connectivity.

**Reply by ObsisMc** [reply]
Thank you very much, I get it now. After `adj[partptr[i]:partptr[i+1], partptr[j]:partptr[j + 1]]`, I can use `value` attribute to find the edges in my orginal graph.
